Reword confidence messages and add ruleId

josiahsprague opened the pull request:
Resolves #4. Rewords confidence messages and adds `ruleId`. I decided to use the word "confidence" instead of "sureness" both in the messages and internally for clarity and consistency. Messages now look like this;
```
5:1-5:68  warning    Hard to read sentence, low confidence           hard-to-read
6:1-6:86  warning    Hard to read sentence, moderate confidence      hard-to-read
7:1-7:98  warning    Hard to read sentence, high confidence          hard-to-read
```

josiahsprague commented:
Oops, I need to fix tests and update the README. :)

wooorm commented on index.js:
```diff
@@ -102,17 +102,21 @@ function report(file, node, threshold, target, results) {
+        );
 
-        message.level = level.toLowerCase();
+        message.level = confidence;
```
@josiahsprague How about adding a new property here, `confidence`, and settings `result` to that? That way, the confidence factor is also available to userland.

josiahsprague replied:
Is the `level` property used anywhere else?

wooorm replied:
Nope!

josiahsprague replied:
Cool, so should I rename it to `confidence`, or keep both? Since the text returned here is changing, this might technically be a breaking change, unless we put in some lines to return the `level` with the same text as before.

wooorm replied:
Like this?

```diff
-        message.level = level.toLowerCase();
+        message.level = confidence;
+        message.confidence = result;
```

(in that case maybe rename the `confidence` variable, not the property, to `label` or something, and `result` to `confidence`?)

I’m OK with doing a major release 👍

josiahsprague replied:
For better consistency, I renamed `confidence` to `confidenceLevel` and `result` to `confidence`, then did;
```
message.level = confidenceLevel;
message.confidence = confidence;
```

How's that look?

wooorm replied:
So `level` is the `string`, and `confidence` the `number`? Maybe, in that case, we could make it even more clear by:

a) replacing `level` (prop) with `confidenceLevel`?
b) replacing `level` (prop) with `label` and `confidenceLevel` (var) with `label`?

I’m leaning towards b, thoughts?

josiahsprague replied:
I'm leaning towards a, because `label` is ambiguous outside of the context of `report()`.

wooorm replied:
OK!

wooorm commented on index.js:
```diff
@@ -102,17 +102,21 @@ function report(file, node, threshold, target, results) {
+        message = file.warn(
+            'Hard to read sentence, ' + confidence + ' confidence',
+            node,
+            'hard-to-read'
```
Could you change `hard-to-read` > `retext-readability`? One less magic string, and it makes more sense to me.

wooorm commented on index.js:
```diff
@@ -102,17 +102,21 @@ function report(file, node, threshold, target, results) {
 
-        message = file.warn(level + ' hard to read sentence', node);
+        message = file.warn(
+            'Hard to read sentence, ' + confidence + ' confidence',
```
How about a format more like: `Hard to read sentence (confidence: ' + confidence + ')'`, which’ll get: `Hard to read sentence (confidence: high)'`

josiahsprague commented:
Does this look good to merge now?

wooorm commented:
Sorry for the mixed signals, but two more Qs 😐

1. How about `confidenceLabel` instead of `confidenceLevel`, as a property? It’s just a bit confusing looking at those properties what they mean, `confidenceLabel` clearly implies it’s a label, so they other one must be the actual raw value... Thoughts?

2. I slightly don’t like the wording of the lowest tier confidence. “Low” sounds a bit too... low?! So, instead of `high`, `moderate`, `low`; maybe `very high`, `high`, `moderate`?

wooorm commented:
Or, `certain`, `likely`, `probable`? https://en.wikipedia.org/wiki/Words_of_estimative_probability

josiahsprague commented:
I think `certain`, `likely`, `probable` still sound a little ambiguous. `very high`, `high`, `moderate` seem more clearer. Or we could just do away with the label altogether, and calculate the message description as a percentage like this;
```
'Hard to read sentence (confidence: ' + (confidence*100).toFixed() + '%)'
// Hard to read sentence (confidence: 71%)
```

wooorm commented:
As those numbers are a bit random, how about `Hard to read sentence (confidence: 5/7)` and the like?

josiahsprague commented:
Displaying the numbers as fractions might be a little bit more difficult, since JS is doing the division automatically. I think displaying them as percentages is intuitive. Right now, four percentages could be displayed, `100%` (7/7), `86%` (6/7), `71%` (5/7), and `57%` (4/7). If the user passes a lower threshold option in, lower percentages could be displayed as well. The advantage of using the percentage is that it gives people an intuitive idea of the meaning without having to understand and remember that there are 7 different formulas being used. And if another formula is added (or removed) in the future, the code can dynamically adapt to that.

wooorm commented:
@josiahsprague Well, we count the succeeding algorithms right? Just before `confidence /= length`, `confidence` is number `0..7`! Something like this:

```js
'Hard to read sentence (confidence: ' + success + '/7)'
```


...where `succes` is `confidence` just before `confidence` is modified?

wooorm commented:
Tip: use 'npm run test' locally, to check if code style is OK!

josiahsprague commented:
Just noticed that. :)

josiahsprague commented:
So I refactored some of the code to make it simpler. Since we're just labeling with the confidence string now, I removed the constants for the separate levels. I also changed the code used to get the count of failed formulas to use `Array.prototype.filter()` so that it's more concise. And I'm just adding `message.confidence` to the message object as a string that looks like this; `'5/7'`.

How's that look?

wooorm commented:
Done! Thanks so much, @josiahsprague, lovely to have had your help!

I did revert the `filter` to a `while` loop, though, I really dig those :)

josiahsprague commented:
Sounds good. :) Will you be releasing updates to npm soon?

wooorm commented:
Yup, in a sec

wooorm commented:
Released!
